File: fetch-data.py
import json
from datetime import timedelta
from glob import glob
from os.path import basename

import requests
import requests_cache
from rich.progress import track

requests_cache.install_cache(
    backend=requests_cache.FileCache(
        user_cache_dir=True,
    ),
    expire_after=timedelta(10),
)


def run():
    for fname in glob("content/*/*.data.json"):
        rich_data = _process_file(fname)
        open(
            fname.replace(".data.", ".data-rich."),
            "w",
        ).write(json.dumps(rich_data, indent=2, ensure_ascii=False))


def _process_file(fname: str):
    data = json.loads(open(fname, encoding="utf-8").read())
    return {
        key: _enrich_data(items)
        for key, items in track(
            data.items(),
            description=basename(fname),
        )
    }


def _enrich_data(items):
    return [_fetch_additional_data(i) for i in items]


def _fetch_additional_data(item: dict[str, str]):
    return item | {
        "github": _fetch_gh_data(item),
        "npmjs": _fetch_npmjs_data(item),
    }


def _fetch_gh_data(item):
    name = item["url"][19:]
    url = f"https://api.github.com/search/repositories?q=@{name}"

    data = requests.get(url).json()

    if not data.get("items"):
        print(f"Repo {name} not found", data)
        return None

    item_data = data["items"][0]
    return {
        "stars": item_data["stargazers_count"],
        "pushedAt": item_data["pushed_at"],
    }


def _fetch_npmjs_data(item):
    name = item.get("npmName") or item.get("name")
    url = f"https://registry.npmjs.org/{name}"

    data = requests.get(url).json()

    releases = list(data["time"].items())
    last_release, last_release_at = releases[-1]

    return {
        "lastRelease": last_release,
        "lastReleaseAt": last_release_at,
    }


if __name__ == "__main__":
    run()

File: fetch-data.ts
/// <reference types="node" />
// https://vitepress.dev/guide/data-loading
import fs from "node:fs/promises"

import {glob} from "glob"
import sqliteStore from "cache-manager-sqlite"
import cacheManager from "cache-manager"
import {Octokit} from "octokit"

import {PlainItem, JsonDataFile, RichItem} from "./content/.vitepress/types"

const octokit = new Octokit({ auth: process.env.GH_TOKEN })
const cache = cacheManager.caching({
  store: sqliteStore,
  name: "packages",
  path: "/tmp/cache.db",
  ttl: 864000 * 1000, // Ten days in milliseconds.
});

// https://stackoverflow.com/questions/9781218/how-to-change-node-jss-console-font-color
const colors = {
  Reset: "\x1b[0m",
  FgRed: "\x1b[31m",
  FgGreen: "\x1b[32m",
};

(async function main() {
  const inputFiles = await glob("content/*/*.data.json")
  for (const fname of inputFiles) {
    await handleFile(fname)
  }
})()

/**
 * Fetch additional data for repositories.
 */
const handleFile = async (fname: string) => {
  console.log(`Processing ${fname}…`)

  const src = await fs.readFile(fname, {encoding: 'utf-8'})
  const srcData: JsonDataFile = JSON.parse(src)

  const total = Object.values(srcData).map(l => l.length).reduce((c, a) => c + a, 0)
  const okSign = `${colors.FgGreen}✓${colors.Reset}`
  const errorSign = `${colors.FgRed}𐄂${colors.Reset}`
  var currentItem = 0

  async function handleItem(item: PlainItem) {
    const wrap = (p: Promise<any>, sourceName: string) => {
      return p.then((r) => {
        console.info(` ${sourceName} ${okSign}`)
        return r
      }).catch((err) => {
        console.error(` ${sourceName} ${errorSign}: ${err}`)
        return undefined
      })
    }
    currentItem++

    console.log(`[${currentItem}/${total}] ${item.name}…`)

    const [github, npmjs] = await Promise.all([
      wrap(fetchGithubData(item), 'github'),
      wrap(fetchNpmjsData(item), 'npmjs'),
    ])


    return {
      ...item,
      github,
      npmjs,
    }
  }

  const richData: Record<string, RichItem[]> = {}

  // Not using parallel processing here to stay below api limits.
  for (const [key, items] of Object.entries(srcData)) {
    const richItems: RichItem[] = []
    for (const item of items) {
      richItems.push(await handleItem(item))
    }
    richData[key] = richItems
  }

  await fs.writeFile(
    fname,
    JSON.stringify(richData, null, 2) + "\n",
    {encoding: 'utf-8'},
  )
};

/**
 * Fetch package data from Github search API
 *
 * For authenticated requests, you can make up to 30 requests per minute.
 * Retries are handled by Octokit automatically.
 *
 * https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28
 */
async function fetchGithubData(item: PlainItem) {
  if (!item.url.startsWith('https://github.com/')) {
    console.log("Skip non-github url.")
    return undefined
  }
  const name = item.url.slice(19);
  const query = `@${name}`;


  const {data: {items}} = await searchGithub(query).catch(error => {
    return {data: {}}
  })

  if (!items?.length) {
    throw new Error(`Repo ${name} not found.`)
  }

  const data: {
    stargazers_count: number;
    pushed_at: string;
  } = items[0];
  return {
    stars: data.stargazers_count,
    pushedAt: data.pushed_at,
  };
}

/**
 * Fetch package data from npm registry
 *
 * API limits - Unknown
 * https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md
 */
async function fetchNpmjsData(item: PlainItem) {
  const name = item.npmName || item.name
  const url = `https://registry.npmjs.org/${name}`
  const {time, error}: {
    time?: Record<string, string>,
    error?: string
  } = await fetchJson(url);
  if (error) {
    throw new Error(error)
  }
  const releases = Object.entries(time);
  const [lastRelease, lastReleaseAt] = releases[releases.length - 1];
  return {
    lastRelease,
    lastReleaseAt,
  };
}

function searchGithub(q: string) {
  return new Promise<any>((resolve, reject) => {
    cache.wrap(
      q,
      (cb) =>
        octokit.rest.search.repos({q})
          .then((r) => cb(null, r))
          .catch((err) => cb(err)),
      {},
      (err, res) => {
        if (err) reject(err);
        else resolve(res);
      },
    );
  });
}

function fetchJson(url: string | URL | Request, init?: RequestInit) {
  // cache-manager v4 uses this awful syntax, but it works!
  // we're using v4, because cache-manager-sqlite still not updated for v5
  // After this is done, we can update and switch to async/await.
  // https://github.com/maxpert/node-cache-manager-sqlite/issues/4
  return new Promise<any>((resolve, reject) => {
    cache.wrap(
      url,
      (cb) =>
        _fetchJson(url, init)
          .then((r) => cb(null, r))
          .catch((err) => cb(err)),
      {},
      (err, res) => {
        if (err) reject(err);
        else resolve(res);
      },
    );
  });
}

async function _fetchJson(url: string | URL | Request, init?: RequestInit) {
  return await (await fetch(url, init)).json();
}
